File: Civitas_Ruby/lib/tablero.rb
#encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'casilla.rb'

module Civitas
  class Tablero
    
    attr_reader :numCasillaCarcel
    
    def initialize(n)
      
      if n >= 1 then
        @numCasillaCarcel = n;
      else
        @numCasillaCarcel = 1;
      end
      
      @casillas = Array.new(0,nil)
      casilla = Casilla.new("Salida")
      @casillas.push(casilla)
      @porSalida = 0
      @tieneJuez = false
      
    end
    
    def correcto(numCasilla = 0)
      if @casillas.count() > @numCasillaCarcel && @tieneJuez && numCasilla >= 0 && numCasilla < @casillas.count() then
        return true
      else
        return false
      end
    end
    
    def getPorSalida()
      if @porSalida > 0 then
        @porSalida -= 1
        return @porSalida+1
      else
        return @porSalida
      end
    end
    
    def añadeCasilla(casilla)
      if @casillas.count() == @numCasillaCarcel then
        carcel = Casilla.newCarcel("Cárcel", @numCasillaCarcel)
        @casillas.push(carcel)
        @casillas.push(casilla)
      else
        @casillas.push(casilla)
        if @casillas.count() == @numCasillaCarcel then
          carcel = Casilla.new("Cárcel")
          @casillas.push(carcel)
        end
      end
      return 0
    end
    
    def añadeJuez()
      if !@tieneJuez then
        juez = Casilla_Juez.new(@numCasillaCarcel,"Juez")
        @casillas.push(juez)
        @tieneJuez = true
      end
      
      return 0
    end
    
    def getCasilla(numCasilla)
      if correcto(numCasilla) then
        return @casillas[numCasilla]
      else
        return nil
      end
    end
    
    def nuevaPosicion(actual, tirada)
      if !correcto() then
        return -1
      elsif  actual + tirada < @casillas.count() then
        return actual + tirada
      else
        @porSalida += 1
        return (actual + tirada) % @casillas.count()
      end
    end
    
    def calcularTirada(origen, destino)
      if (destino - origen) < 0 then
        return destino - origen + @casillas.count()
      else
        return destino - origen
      end
    end
    
    private :correcto
  end
end

File: Civitas_TUI/src/civitas/Tablero.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.ArrayList;

/**
 *
 * @author alejandrof
 */
public class Tablero {
   private int numCasillaCarcel;
   private int porSalida;
   private Boolean tieneJuez;
   ArrayList<Casilla> casillas;
   
   public Tablero(int carcel){
       if(carcel >= 1)
           numCasillaCarcel = carcel;
       else
           numCasillaCarcel = 1;
       
       Casilla salida = new Casilla("Salida");
       
       casillas = new ArrayList<>();
       casillas.add(salida);
       
       porSalida = 0;
       tieneJuez = false;
   }
   
   private Boolean correcto(){
       return (casillas.size() > numCasillaCarcel) && tieneJuez;
   }
   
   private Boolean correcto(int numCasilla){
       return this.correcto() && (numCasilla < casillas.size());
   }
   
   int getCarcel(){
       return numCasillaCarcel;
   }
   
   int getPorSalida(){
       if(porSalida > 0){
           porSalida--;
           return porSalida+1;
       }
       else
           return porSalida;
   }
   
   void añadeCasilla(Casilla casilla){
       if(casillas.size() == numCasillaCarcel){
           Casilla carcel = new Casilla("Cárcel");
           casillas.add(carcel);
           casillas.add(casilla);
       }
       else{
           casillas.add(casilla);
           if(casillas.size() == numCasillaCarcel){
               Casilla carcel = new Casilla("Cárcel");
               casillas.add(carcel);
           }
       }
   }
   
   void añadeJuez(){
       if(!tieneJuez){
           Casilla juez = new Casilla_Juez(numCasillaCarcel,"Juez");
           this.añadeCasilla(juez);  
           tieneJuez = true;
       }
   }
   
   Casilla getCasilla(int numCasilla){
       if(this.correcto(numCasilla))
           return casillas.get(numCasilla);
       else
           return null;
   }
   
   int nuevaPosicion(int actual, int tirada){
       if(!this.correcto())
           return -1;
       else{
           if(actual+tirada < casillas.size()){
               return actual+tirada;
           }
           
           else{
               porSalida++;
               return (actual+tirada) % casillas.size();
           }
       }
   }
   
   int calcularTirada(int origen, int destino){
       if((destino - origen) < 0)
           return destino-origen+casillas.size();
       else
           return destino-origen;
   }
}
